resources/src/asset.rs

```diff
@@ -45,33 +169,89 @@ struct Flag {
 /// An asset consists of an type, code, and issuer.
 /// Any asset can be traded for any other asset.
 /// https://www.stellar.org/developers/horizon/reference/resources/asset.html
-#[derive(Serialize, Deserialize, Debug)]
+#[derive(Debug)]
 pub struct Asset {
+    asset_identifier: AssetIdentifier,
+    amount: Amount,
+    num_accounts: u32,
+    flags: Flag,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct IntermediateAsset {
     asset_type: String,
-    asset_code: String,
-    asset_issuer: String,
+    #[serde(skip_serializing_if = "Option::is_none")] asset_code: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")] asset_issuer: Option<String>,
     amount: Amount,
     num_accounts: u32,
     flags: Flag,
 }
 
+impl<'de> Deserialize<'de> for Asset {
+    fn deserialize<D>(d: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let rep: IntermediateAsset = IntermediateAsset::deserialize(d)?;
+        let asset_identifier: Result<AssetIdentifier, D::Error> =
+            AssetIdentifier::new(rep.asset_type, rep.asset_code, rep.asset_issuer).map_err(|_| {
+                de::Error::custom("Code and issuer are required for non-native assets")
```

Comment:
This assumes to know what the error is when it could be any `String`. We should probably match on an explicit error type.